ui/src/shared/apis/queryCache.ts

```diff
@@ -0,0 +1,164 @@
+        this.resetCacheByID(id)
+      }
+    }
+  }
+
+  getFromCache = (id: string, variableID: string): RunQueryResult | null => {
+    // no existing query match
+    if (!this.cache[id]) {
+      return null
+    }
+    // query match with no existing variable match
+    if (this.cache[id].variables !== variableID) {
+      this.resetCacheByID(id)
+      return null
+    }
+    // query & variable match with an expired result
+    if (Date.now() - this.cache[id].dateSet > TIME_INVALIDATION) {
+      this.resetCacheByID(id)
+      return null
+    }
+    return this.cache[id].values
+  }
+
+  resetCacheByID = (id: string): void => {
+    if (!this.cache[id]) {
+      return
+    }
+    delete this.cache[id]
+  }
+
+  resetCache = (): void => {
+    this.cache = {}
+  }
+
+  setCacheByID = (
+    queryID: string,
+    variableID: string,
+    values: RunQueryResult,
+    isCustomTime: boolean = false
+  ): void => {
+    this.cache[queryID] = {
+      dateSet: Date.now(),
+      isCustomTime,
+      values,
+      variables: variableID,
+    }
+  }
+
+  startWatchDog = () => {
+    setInterval(() => {
+      this._cleanExpiredQueries()
+    }, TIME_INVALIDATION / 2)
+
+    this._cleanExpiredQueries()
+  }
+}
+
+const queryCache = new QueryCache()
+// Set an interval to check for expired data to invalidate
+queryCache.startWatchDog()
+
+export const resetQueryCache = (): void => {
+  queryCache.resetCache()
+}
+
+export const resetQueryCacheByQuery = (query: string): void => {
+  const queryID = `${hashCode(query)}`
+  queryCache.resetCacheByID(queryID)
+}
+
+export const getRunQueryResults = (
+  orgID: string,
+  query: string,
+  state: AppState,
+  abortController?: AbortController
+): CancelBox<RunQueryResult> => {
+  const usedVars = filterUnusedVarsBasedOnQuery(getAllVariables(state), [query])
+  const variables = sortBy(usedVars, ['name'])
+  const simplifiedVariables = variables.map(v => asSimplyKeyValueVariables(v))
+  const stringifiedVars = JSON.stringify(simplifiedVariables)
+  // create the queryID based on the query & vars
+  const queryID = `${hashCode(query)}`
+  const variableID = `${hashCode(stringifiedVars)}`
+
+  const cacheResults: RunQueryResult | null = queryCache.getFromCache(
+    queryID,
+    variableID
+  )
+
+  // check the cache based on text & vars
+  if (cacheResults) {
+    const controller = abortController || new AbortController()
+    return {
+      promise: new Promise(resolve => resolve(cacheResults)),
+      cancel: () => controller.abort(),
+    }
+  }
+  const variableAssignments = variables
+    .map(v => asAssignment(v))
+    .filter(v => !!v)
+  // otherwise query & set results
+  const extern = buildVarsOption(variableAssignments)
+  const results = runQuery(orgID, query, extern, abortController)
+  results.promise = results.promise.then(res => {
+    // if the timeRange is non-relative (i.e. a custom timeRange || the query text has a set time range)
```

Comment:
```suggestion
    // if the timeRange is non-relative (i.e. a custom timeRange or the query text has a set time range)
```
don't mix logical operators in comments. it makes all the `&` you have in your comments look like a bitwise operator.